jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -1041,10 +1070,18 @@ public void onFrame(Frame frame)
 
     protected void onStreamOpened(IStream stream)
     {
+        streamCount.addAndGetLo(1);
     }
 
     protected void onStreamClosed(IStream stream)
     {
+        if (streamCount.addAndGetLo(-1) == 0)
+        {
+            Callback.Completable callback = closeCallback;
+            // Only send the close frame if we can flip Hi, see shutdown().
+            if (callback != null && streamCount.compareAndSet(0, 1, 0, 0))
```

Comment:
I don't like having two CaS operations serially.  This one should be a loop something like:
```java
Callback callback;
while(true)
{
    long count = streamCount.get();
    int streams = AtomicBiInteger.getLo(count) - 1;
    int close = AtomicBiInteger.getHi(count);
    callback =   (streams == 0 && close == 1) ? closeCallback : null;
    if (streamCount.compareAndSet(count, 0, streams))
        break;
}
control(null, callback, closeFrame);
```

Reply:
Done, actually in a different way because the snippet above is wrong, but I implemented a single CAS.